posts/inside-rust/2023-05-01-cargo-postmortem.md

```diff
@@ -0,0 +1,29 @@
+
+Following this incident, the Cargo team went through a postmortem analysis, as is appropriate for incidents with (a potential for) broad blast-radius or otherwise significant impact. This time around, we followed a particular structured postmortem template in the hopes that it might make the resulting write-up more thorough, insightful, and actionable, and we ultimately found that it gave us a better understanding of the underlying root causes and failing/missing safe-guards. So, we wanted to share our experience with other Rust teams in case they may find it similarly useful, either in part or as a whole. 
+
+The postmortem template consists of four parts:
+
+* **What happened:** a summary that provides context for the event, including metrics or graphs that illustrate the impact of the event if available. This should include a summary of any user-facing impacts or experience during the event.
+* **How we responded:** a timeline that describes all the events that occurred during the incident including specific date/time to the extent that they are known, as well as answers to the following four questions:
+    * How was the event detected?
+    * How could time to detection be improved?
+    * How did you reach the point where you knew how to mitigate the impact?
+    * How could time to mitigation be improved?
+* **Why the event happened:** this is the juicy part. Here, we use the [Five Whys] approach to dig deeply down until the incident’s root causes are identified. Each answer is meant to spawn one or more why questions, until you’re confident that the left answers are fundamentally root causes. It’s also worth pointing out explicitly that “operator error” is **never** a root cause, and that this is not a process for assigning blame. Instead, any operator error is a symptom of a missing or broken mechanism, and the answers should focus on identifying those inadequate mechanisms.
+* **How to fix it:** The outcome of the Five Whys exercise is a list of root causes that should be addressed to reduce the risk of a similar incident in the future. From these root causes, we produce short- and medium-term “action items” along with specific owners wherever possible. Long-term solutions can be discussed too, although the focus of action items should be on more immediate mitigation steps that will be taken relatively soon. Each action item is assigned a priority, and is then generally turned into a GitHub issue where applicable. Any items identified as urgent we start working on immediately, while other action items usually fall into the categories “soon” or “once feasible”.
+
+*Note: to ensure the focus stays on mechanisms and processes, not individuals, individuals should not be named unless absolutely necessary. Use terms like "a contributor", "the maintainer", "a libs team member", etc.*
+
+So, without further ado, here is [the postmortem for the aforementioned Cargo incident][postmortem].
```

Comment:
I might recommend inlining this rather than linking to hackmd, so that we avoid possible issues with permissions (it doesn't look like it's globally editable right now, but that can change over time by accident) and save readers a click.

But no strong opinion here.

Reply:
Good point!

We were thinking that the postmortem is not particularly important here. This post aims to share the experience and the template. What about turning it into a Cargo issue on GitHub? That would be safer and more familiar to readers.

Reply:
Updated. The link is now pointing to a GitHub issue